extensions/common/auth/auth-delegated/src/main/java/org/eclipse/edc/api/auth/delegated/JwksPublicKeyResolver.java

```diff
@@ -0,0 +1,116 @@
+import com.nimbusds.jose.jwk.JWKSelector;
+import com.nimbusds.jose.jwk.source.JWKSource;
+import com.nimbusds.jose.jwk.source.JWKSourceBuilder;
+import com.nimbusds.jose.proc.SecurityContext;
+import com.nimbusds.jose.proc.SimpleSecurityContext;
+import org.eclipse.edc.keys.spi.KeyParserRegistry;
+import org.eclipse.edc.keys.spi.PublicKeyResolver;
+import org.eclipse.edc.spi.EdcException;
+import org.eclipse.edc.spi.monitor.Monitor;
+import org.eclipse.edc.spi.result.Result;
+import org.jetbrains.annotations.Nullable;
+
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.security.PublicKey;
+import java.util.List;
+import java.util.Optional;
+
+import static com.nimbusds.jose.jwk.source.JWKSourceBuilder.DEFAULT_CACHE_REFRESH_TIMEOUT;
+import static com.nimbusds.jose.jwk.source.JWKSourceBuilder.DEFAULT_RATE_LIMIT_MIN_INTERVAL;
+import static com.nimbusds.jose.jwk.source.JWKSourceBuilder.DEFAULT_REFRESH_AHEAD_TIME;
+
+/**
+ * A {@link PublicKeyResolver} that resolves a JSON Web Key Set from a URL and parses the JWK with the given ID
+ */
+public class JwksPublicKeyResolver implements PublicKeyResolver {
+    private final Monitor monitor;
+    private final KeyParserRegistry keyParserRegistry;
+    private final JWKSource<SecurityContext> jwkSource;
+
+    /**
+     * Instantiates the resolver
+     *
+     * @param keyParserRegistry Should contain all relevant key parsers. The minimum recommendation is adding a {@code JwkParser}.
+     * @param jwksUrl           The URL of the public key server, where a JWK Set can be obtained.
+     * @param cacheValidityMs   The time in milliseconds that public keys may be cached locally.
+     * @param monitor           A monitor
+     * @throws EdcException if the jwksUrl is malformed
+     */
+    public JwksPublicKeyResolver(KeyParserRegistry keyParserRegistry, String jwksUrl, long cacheValidityMs, Monitor monitor) {
+        this.monitor = monitor;
+        this.keyParserRegistry = keyParserRegistry;
+
+        try {
+            var builder = JWKSourceBuilder.create(URI.create(jwksUrl).toURL()).retrying(false);
+            if (cacheValidityMs > 0) {
+                builder.cache(cacheValidityMs, DEFAULT_CACHE_REFRESH_TIMEOUT);
+
+                // rate-limit must be < cache TTL, this would cause the cache to be refreshed more often than allowed
+                if (cacheValidityMs < DEFAULT_RATE_LIMIT_MIN_INTERVAL) {
+                    builder.rateLimited(cacheValidityMs - 1);
+                }
+                // cache TTL must be > refresh-ahead time plus refresh timeout
+                if (cacheValidityMs < DEFAULT_REFRESH_AHEAD_TIME + DEFAULT_CACHE_REFRESH_TIMEOUT) {
+                    builder.refreshAheadCache(false);
+                }
+
+            }
+            jwkSource = builder.build();
+        } catch (MalformedURLException e) {
+            monitor.warning("Malformed JWK URL: " + jwksUrl, e);
+            throw new EdcException(e);
+        }
```

Comment:
looks like this would fit better in a factory method and have the `JWKSource` passed already built to the constructor

Reply:
that would leak implementation details (`JWKSource`) outside of the class, and it would put the need to know the intrinsics of the `JWKSource` on the caller.

Reply:
the factory method can actually stay in the class. I'm always a bit reclutant to add logic into a constructor, plus, as a rule of thumb, it's always good to know that a constructor cannot fail, leaving the dirty work to factories/builders

Reply:
ah, got you. yeah, that sounds good